revocations.c

```diff
@@ -0,0 +1,528 @@
+	}
+
+	// Calculate and reserve room for the bitmap size.
+	r = grl->parsed;
+	while (r->next != NULL) {
+		r = r->next;
+
+		// Add the serial to the 1st section.
+		if ((ret = sshbuf_put_u64(blob, r->serial)) < 0) {
+			debug3("hibagrl_encode: sshbuf_put_u64 returned %d: %s", ret, ssh_err(ret));
+			return HIBA_INTERNAL_ERROR;
+		}
+		if ((ret = sshbuf_put_u64(blob, offset)) < 0) {
+			debug3("hibagrl_encode: sshbuf_put_u64 returned %d: %s", ret, ssh_err(ret));
+			return HIBA_INTERNAL_ERROR;
+		}
+
+		offset += r->size;
+	}
+
+	if ((ret = sshbuf_put_u64(blob, offset)) < 0) {
+		debug3("hibagrl_encode: sshbuf_put_u64 returned %d: %s", ret, ssh_err(ret));
+		return HIBA_INTERNAL_ERROR;
+	}
+	if ((ret = sshbuf_allocate(blob, offset)) < 0) {
+		debug3("hibagrl_encode: sshbuf_allocate(%" PRIu64 ") returned %d: %s",
+		       offset, ret, ssh_err(ret));
+		return HIBA_INTERNAL_ERROR;
+	}
+
+	debug3("hibagrl_encode: bitmap section @%zu", sshbuf_len(blob));
+
+	r = grl->parsed;
+	while (r->next != NULL) {
+		int m;
+		r = r->next;
+
+		for (m = 0; m < r->size; ++m) {
+			if ((ret = sshbuf_put_u8(blob, r->map[m])) < 0) {
+				debug3("hibagrl_encode: sshbuf_put_u8 returned %d: %s", ret, ssh_err(ret));
+				return HIBA_INTERNAL_ERROR;
+			}
+		}
+		offset += r->size + sizeof(u_int16_t);
+
+		++i;
+	}
+
+	debug2("hibagrl_encode: patch bitmap section size = %" PRIu64, offset);
+
+	// Write trailer magic.
+	if ((ret = sshbuf_put_u32(blob, HIBA_GRL_MAGIC)) < 0) {
+		debug3("hibagrl_encode: sshbuf_put_u32 returned %d: %s", ret, ssh_err(ret));
+		return HIBA_INTERNAL_ERROR;
+	}
+
+	return HIBA_OK;
+}
+
+#define _BLOCK_FOR_ID(id) ((id) >> 3)
+#define _OFFSET_FOR_ID(id) ((id) & 0x7)
+
+// We want to update the offset in the right block.
+#define REVOKE_ID(r, id) do { \
+		debug("REVOKE_ID(%d): map[%d] |= %02x", id, _BLOCK_FOR_ID(id), _OFFSET_FOR_ID(id)); \
+		((r)->map[_BLOCK_FOR_ID(id)] |= 1 << _OFFSET_FOR_ID(id)); \
+	} while (0)
+
+int
+hibagrl_revoke_grant(struct hibagrl *grl, u_int64_t serial, u_int32_t lo, u_int32_t hi) {
+	u_int32_t i;
+	int required_size = 0;
+	struct parsed_revocations *r;
+
+	if (grl == NULL)
+		return HIBA_BAD_PARAMS;
+	if (grl->parsed == NULL)
+		return HIBA_BAD_PARAMS;
+
+	// Look for where to add the revocation.
+	r = grl->parsed;
+	while (r->next) {
+		if (r->next->serial >= serial)
+			break;
+		r = r->next;
+	}
+
+	// Maybe add a new serial to the ordered list.
+	if ((r->next == NULL) || (r->next->serial > serial)) {
+		struct parsed_revocations *next = r->next;
+		debug2("hibagrl_revoke_grant: adding a new entry for serial 0x%" PRIx64 " after 0x%" PRIx64 , serial, r->serial);
+		r->next = calloc(1, sizeof(struct parsed_revocations));
+		r = r->next;
+		r->next = next;
+		r->serial = serial;
+
+		++grl->n_serials;
+	} else {
+		r = r->next;
+		debug2("hibagrl_revoke_grant: reusing existing entry for serial 0x%" PRIx64, serial);
+	}
+
+	required_size = _BLOCK_FOR_ID(hi) + 1;
+
+	// Allocate or update the bitmap
```

Comment:
If the revocation was inserted, rather than appended, wouldn't the bitmap indices be off now?

Reply:
The block and offet for a grant index are absolute, and the bitmap is grown to the size corresponding to the highest grant index it contains (if revoking 1 grant at position 1024, the bitmap will be 128 bytes large)